tokio/src/sync/watch.rs

```diff
@@ -610,6 +610,93 @@ impl<T> Receiver<T> {
+    /// the channel. Once the closure returns `true`, this method will return a
+    /// reference to the value that was passed to the closure.
+    ///
+    /// Before `wait_for` starts waiting for changes, it will call the closure
+    /// on the current value. If the closure returns `true` when given the
+    /// current value, then `wait_for` will immediately return a reference to
+    /// the current value. This is the case even if the current value is already
+    /// considered seen.
+    ///
+    /// The watch channel only keeps track of the most recent value, so if
+    /// several messages are sent faster than `wait_for` is able to call the
+    /// closure, then it may skip some updates. Whenever the closure is called,
+    /// it will be called with the most recent value.
+    ///
+    /// When this function returns, the value that was passed to the closure
+    /// when it returned `true` will be considered seen.
+    ///
+    /// If the channel is closed, then `wait_for` will return a `RecvError`.
+    /// Once this happens, no more messages can ever be sent on the channel.
+    /// When an error is returned, it is guaranteed that the closure has been
+    /// called on the last value, and that it returned `false` for that value.
+    /// (If the closure returned `true`, then the last value would have been
+    /// returned instead of the error.)
+    ///
+    /// Like the `borrow` method, the returned borrow holds a read lock on the
+    /// inner value. This means that long-lived borrows could cause the producer
+    /// half to block. It is recommended to keep the borrow as short-lived as
+    /// possible. See the documentation of `borrow` for more information on
+    /// this.
+    ///
+    /// [`Receiver::changed()`]: crate::sync::watch::Receiver::changed
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio::sync::watch;
+    ///
+    /// #[tokio::main]
+    ///
+    /// async fn main() {
+    ///     let (tx, _rx) = watch::channel("hello");
+    ///
+    ///     tx.send("goodbye").unwrap();
+    ///     
+    ///     // here we subscribe to a second receiver
+    ///     // now in case of using `changed` we would have
+    ///     // to first check the current value and then wait
+    ///     // for changes or else `changed` would hang.
+    ///     let mut rx2 = tx.subscribe();
+    ///     
+    ///     // in place of changed we have use `wait_for`
+    ///     // which would automatically check the current value
+    ///     // and wait for changes until the closure returns true.
+    ///     assert!(rx2.wait_for(|val| *val == "goodbye").await.is_ok());
+    ///     assert_eq!(*rx2.borrow(), "goodbye");
+    /// }
+    /// ```
+    pub async fn wait_for(
+        &mut self,
+        mut f: impl FnMut(&T) -> bool,
+    ) -> Result<Ref<'_, T>, error::RecvError> {
+        loop {
+            if f(&self.borrow_and_update()) {
+                return Ok(self.borrow());
+            }
```

Comment:
I didn't notice this initially, but this is wrong. You need to reuse the `Ref` object.
```Rust
{
    let val = self.borrow_and_update();
    if f(&val) {
        return Ok(val);
    }
}
```
When you lock the `RwLock` twice like this, you risk that someone changed the value between the call to `self.borrow_and_update()` and `self.borrow()`. That would be wrong since we promise to return the same value that we passed to the closure.

It's possible that you're lucky here and that the `borrow_and_update` lock guard is still alive by the time you call `self.borrow()`, meaning that a write lock could not have been taken, preventing modifications. However, in that case, you can end up with a deadlock because a writer might be in front of the `self.borrow()` call in the lock queue.

Reply:
Hey! I keep trying to do this 

```diff
diff --git a/tokio/src/sync/watch.rs b/tokio/src/sync/watch.rs
index dbd3fd2e..55fb461d 100644
--- a/tokio/src/sync/watch.rs
+++ b/tokio/src/sync/watch.rs
@@ -675,8 +675,9 @@ impl<T> Receiver<T> {
         mut f: impl FnMut(&T) -> bool,
     ) -> Result<Ref<'_, T>, error::RecvError> {
         loop {
-            if f(&self.borrow_and_update()) {
-                return Ok(self.borrow());
+            let val = self.borrow_and_update();
+            if f(&val) {
+                return Ok(val);
             }
```
and i get the borrow checker complaint

```rust
error[E0499]: cannot borrow `*self` as mutable more than once at a time
   --> tokio/src/sync/watch.rs:678:23
    |
674 |         &mut self,
    |         - let's call the lifetime of this reference `'1`
...
678 |             let val = self.borrow_and_update();
    |                       ^^^^^^^^^^^^^^^^^^^^^^^^ `*self` was mutably borrowed here in the previous iteration of the loop
679 |             if f(&val) {
680 |                 return Ok(val);
    |                        ------- returning this value requires that `*self` is borrowed for `'1`
```

I cant really understand why? val is not being used anywhere else so it should be dropped no and release the lock on self? could you help 😅😅😅😓

Reply:
Ah ... That's a case where the borrow-checker is too strict. If you search for polonious, then you should find similar discussions.

In this particular case, the issue would probably go away if you inserted the definition of `borrow_and_update` verbatim, because then the returned reference only takes an immutable borrow rather than a mutable one.

I can provide more help later.

Reply:
> the issue would probably go away if you inserted the definition of borrow_and_update verbatim

didn't understand inserting the definition verbatim meaning?

> I can provide more help later.

No problem at all 

I am trying to search alt solutions

Thank You!

Reply:
oh ok i think you mean inserting the definition of borrow and update inside the func

Reply:
ok i tried an initial attempt to add the borrow_and_update definition inside the function but not much success yet

Reply:
The following patch should work. You can apply it with `git am`.
```patch
From 613faad0ca9382f1d2314f6a1e166a26976ed6b0 Mon Sep 17 00:00:00 2001
From: Alice Ryhl <aliceryhl@google.com>
Date: Sun, 23 Apr 2023 17:13:32 +0200
Subject: [PATCH] inline borrow_and_update into wait_for

---
 tokio/src/sync/watch.rs | 65 +++++++++++++++++++++++------------------
 1 file changed, 36 insertions(+), 29 deletions(-)

diff --git a/tokio/src/sync/watch.rs b/tokio/src/sync/watch.rs
index dbd3fd2e..3dc19231 100644
--- a/tokio/src/sync/watch.rs
+++ b/tokio/src/sync/watch.rs
@@ -595,19 +595,7 @@ impl<T> Receiver<T> {
     /// }
     /// ```
     pub async fn changed(&mut self) -> Result<(), error::RecvError> {
-        loop {
-            // In order to avoid a race condition, we first request a notification,
-            // **then** check the current value's version. If a new version exists,
-            // the notification request is dropped.
-            let notified = self.shared.notify_rx.notified();
-
-            if let Some(ret) = maybe_changed(&self.shared, &mut self.version) {
-                return ret;
-            }
-
-            notified.await;
-            // loop around again in case the wake-up was spurious
-        }
+        changed_impl(&self.shared, &mut self.version).await
     }
 
     /// Waits for a value that satisifes the provided condition.
@@ -656,13 +644,13 @@ impl<T> Receiver<T> {
     ///     let (tx, _rx) = watch::channel("hello");
     ///
     ///     tx.send("goodbye").unwrap();
-    ///     
+    ///
     ///     // here we subscribe to a second receiver
     ///     // now in case of using `changed` we would have
     ///     // to first check the current value and then wait
     ///     // for changes or else `changed` would hang.
     ///     let mut rx2 = tx.subscribe();
-    ///     
+    ///
     ///     // in place of changed we have use `wait_for`
     ///     // which would automatically check the current value
     ///     // and wait for changes until the closure returns true.
@@ -674,25 +662,28 @@ pub async fn wait_for(
         &mut self,
         mut f: impl FnMut(&T) -> bool,
     ) -> Result<Ref<'_, T>, error::RecvError> {
+        let mut closed = false;
         loop {
-            if f(&self.borrow_and_update()) {
-                return Ok(self.borrow());
-            }
+            {
+                let inner = self.shared.value.read().unwrap();
+
+                let new_version = self.shared.state.load().version();
+                let has_changed = self.version != new_version;
+                self.version = new_version;
 
-            match self.changed().await {
-                Ok(_) => {}
-                Err(e) => {
-                    let updated_val = self.borrow_and_update();
-                    // check if the version has changed since the last time we
-                    // called the closure.
-                    if updated_val.has_changed && f(&updated_val) {
-                        return Ok(updated_val);
+                if !closed || has_changed {
+                    if f(&inner) {
+                        return Ok(Ref { inner, has_changed });
                     }
-                    // if the version hasn't changed, then the channel has been
-                    // closed. We return the error.
-                    return Err(e);
                 }
             }
+
+            if closed {
+                return Err(error::RecvError(()));
+            }
+
+            // Wait for the value to change.
+            closed = changed_impl(&self.shared, &mut self.version).await.is_err();
         }
     }
 
@@ -741,6 +732,22 @@ fn maybe_changed<T>(
     None
 }
 
+async fn changed_impl<T>(shared: &Shared<T>, version: &mut Version) -> Result<(), error::RecvError> {
+    loop {
+        // In order to avoid a race condition, we first request a notification,
+        // **then** check the current value's version. If a new version exists,
+        // the notification request is dropped.
+        let notified = shared.notify_rx.notified();
+
+        if let Some(ret) = maybe_changed(shared, version) {
+            return ret;
+        }
+
+        notified.await;
+        // loop around again in case the wake-up was spurious
+    }
+}
+
 impl<T> Clone for Receiver<T> {
     fn clone(&self) -> Self {
         let version = self.version;

base-commit: 09652f8eeca6171feb897aad66b26b44445c6b3e
-- 
2.40.0

```

Reply:
Done!

Reply:
Thanks. Seems like I gave you code with a rustfmt failure. Sorry about that. :)